app/models/metric/targets.rb

```diff
@@ -9,22 +9,23 @@ def self.perf_capture_always=(options)
   end
 
   def self.capture_infra_targets(zone, options)
-    # Preload all of the objects we are going to be inspecting.
-    includes = {:ext_management_systems => {:hosts => {:ems_cluster => :tags, :tags => {}}}}
-    includes[:ext_management_systems][:hosts][:storages] = :tags unless options[:exclude_storages]
-    MiqPreloader.preload(zone, includes)
+    load_infra_targets_data(zone, options)
+    all_hosts = capture_host_targets(zone)
+    targets = enabled_hosts = only_enabled(all_hosts)
+    targets += capture_storage_targets(all_hosts) unless options[:exclude_storages]
+    targets += capture_vm_targets(enabled_hosts) unless options[:exclude_vms]
 
-    targets = zone.hosts
-    targets += zone.storages.select { |s| Storage::SUPPORTED_STORAGE_TYPES.include?(s.store_type) } unless options[:exclude_storages]
+    targets
+  end
 
-    # If it can and does have a cluster, then ask that, otherwise, ask host itself.
-    targets = targets.select do |t|
-      t.respond_to?(:ems_cluster) && t.ems_cluster ? t.ems_cluster.perf_capture_enabled? : t.perf_capture_enabled?
+  # Filter to enabled hosts. If it has a cluster consult that, otherwise consult the host itself.
+  # 
+  # NOTE: if capture_storage takes only enabled, then move
+  # this logic into capture_host_targets
+  def self.only_enabled(hosts)
+    hosts.select do |host|
+      host.ems_cluster ? host.ems_cluster.perf_capture_enabled? : host.perf_capture_enabled?
```

Comment:
but that's not really accurate either, and it feels like something is wrong...even with the original code.

This assumes if a host is on a cluster we use that value only, and that's not true.  In the UI, you can choose to enable a single host, and not the entire cluster.